lib/identity_cache/query_api.rb

```diff
@@ -183,6 +184,7 @@ def get_embedded_association(record, association, options) #:nodoc:
 
       def coder_from_record(record) #:nodoc:
         unless record.nil?
+          record.readonly! if IdentityCache.fetch_read_only_records
```

Comment:
@fbogsany: In one of the previous commit, we were calling `load_and_readonlyify` to mark records as read-only at a higher level such as [`fetch_recursively_cached_association`](https://github.com/Shopify/identity_cache/pull/282/commits/5647e89b96ff6edcb52a6e3c5a2fd759fb875322#diff-17e7d279c636296c9b82ebbef85db611R429). However, the issue was that when calling a `fetch_associated_records` on an item deterministically pulled from the db,  `item.associated_record` was being marked as read-only. The idea here - not ideal though - was to mark them as read-only at a lower level, while preserving the integrity of the records pulled deterministically from the db.

The other alternative that @dylanahsmith suggested was to duplicate records and mark those as read-only.